Web Inspector: Reorganize Web Inspector-scoped device settings

rcaliman-apple opened the pull request:
#### 218530454da476df1d7d2a2c20c5fd58e880b105
<pre>
Web Inspector: Reorganize Web Inspector-scoped device settings
<a href="https://bugs.webkit.org/show_bug.cgi?id=252440">https://bugs.webkit.org/show_bug.cgi?id=252440</a>

Reviewed by Devin Rousso.

Reorganizes the device settings popover to show settings appropriate
to the inspected target. Reuses styles from the popover for user preferences overrides.

Settings are copied verbatim from `Main.js`, organized into groups,
and moved to the new `WI.OverrideDeviceSettingsPopover`. Labels are updated.

A new `WI.DeviceSettingsManager` is introduced to manage the collection of
device settings, user agent, and screen size overrides, previously kept in `Main.js`.
It also holds the responsibility to apply overrides on target initalization.

The methods to set overrides are moved mostly unchanged
from `Main.js` to WI.DeviceSettingsManager`:

```
function applyOverriddenUserAgent() -&gt; DeviceSettingsManager.overrideUserAgent()
function applyOverriddenSetting()   -&gt; DeviceSettingsManager.overrideSetting()
function applyOverriddenScreenSize() -&gt; DeviceSettingsManager.overrideScreenSize()
```

A generalized `WI.SettingsPopover` is extracted from `WI.OverrideUserPreferencesPopover`
to handle showing and positioning the popover. The content is provided by subclasses.

A new `WI.OverrideDeviceSettingsPopover` is introduced to show settings
appropriate to the inspected target. More settings are shown in the popover
for remote targets. For local inspection, most of these settings are available
in the develop menu.

The capability to override screen sizes is available only for non-Cocoa platforms.
The code for setting this override is copied verbatim and adapted slightly to
use the new `WI.DeviceSettingsManager`.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Base/Main.js:
(WI.loaded):
(WI.initializeTarget):

Most of the code from `Main.js` has moved to `WI.DeviceSettingsManager`
and `WI.OverrideDeviceSettingsPopover` according to responsibilites (controller or view).

* Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js: Added.
(WI.DeviceSettingsManager):
(WI.DeviceSettingsManager.prototype.initializeTarget):
(WI.DeviceSettingsManager.supportsSetScreenSizeOverride):
(WI.DeviceSettingsManager.prototype.get defaultSettingsChanged):
(WI.DeviceSettingsManager.prototype.get overridenDeviceUserAgent):
(WI.DeviceSettingsManager.prototype.get overridenDeviceScreenSize):
(WI.DeviceSettingsManager.prototype.get overridenDeviceSettings):
(WI.DeviceSettingsManager.prototype.overrideDeviceSetting):
(WI.DeviceSettingsManager.prototype.overrideUserAgent):
(WI.DeviceSettingsManager.prototype.overrideScreenSize):

* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Test.html:
* Source/WebInspectorUI/UserInterface/Views/Main.css:
(.device-settings-content): Deleted.
(.device-settings-content &gt; table): Deleted.
(.device-settings-content &gt; table &gt; tr &gt; td:first-child): Deleted.
(.device-settings-content .container): Deleted.
(.device-settings-content .container &gt; * + *): Deleted.
(.device-settings-content .column): Deleted.
(.device-settings-content .user-agent select): Deleted.
(.device-settings-content .user-agent input): Deleted.
(.device-settings-content label + label): Deleted.
(.device-settings-content label &gt; input): Deleted.
(.device-settings-content &gt; .reference-page-link-container): Deleted.

Removed styles for device settings popover from `Main.css`
and replaced with streamlined styles in `SettingsPopover.css`
and `OverrideDeviceSettingsPopover.css`.

* Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.css: Copied from Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.css.
(.popover .settings-content:is(.device-settings-content) &gt; label):
(.popover .settings-content:is(.device-settings-content) &gt; label &gt; input):
(.popover .settings-content:is(.device-settings-content) &gt; .select-with-input-container):
(.popover .settings-content:is(.device-settings-content) &gt; .reference-page-link-container):
* Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js: Added.
(WI.OverrideDeviceSettingsPopover):
(WI.OverrideDeviceSettingsPopover.prototype.createContentElement):
(WI.OverrideDeviceSettingsPopover.prototype._createSettingsSection):
(WI.OverrideDeviceSettingsPopover.prototype._createUserAgentSection):
(WI.OverrideDeviceSettingsPopover.prototype._createScreenSizeSection):
* Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.css:
(.popover .settings-content:is(.user-preferences-content)):
(.popover .settings-content:is(.user-preferences-content) &gt; h1):
(.popover .settings-content:is(.user-preferences-content) &gt; label):
(.popover .settings-content:is(.user-preferences-content) &gt; select):
(.popover .user-preferences-content): Deleted.
(.popover .user-preferences-content &gt; h1): Deleted.
(.popover .user-preferences-content &gt; h1:not(:first-child)): Deleted.
(.popover .user-preferences-content &gt; label): Deleted.
(.popover .user-preferences-content &gt; select): Deleted.

Generalized styles for settings popovers in `SettingsPopover.css`
and adapted `OverrideDeviceSettingsPopover.css` to build on them.

* Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.js:
(WI.OverrideUserPreferencesPopover):
(WI.OverrideUserPreferencesPopover.prototype.show):
(WI.OverrideUserPreferencesPopover.prototype._createSelectElement):
(WI.OverrideUserPreferencesPopover.prototype._presentOverTargetElement): Deleted.
(WI.OverrideUserPreferencesPopover.prototype._createContentElement): Deleted.
* Source/WebInspectorUI/UserInterface/Views/SettingsPopover.css: Copied from Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.css.
(.popover .settings-content):
(.popover .settings-content &gt; h1):
(.popover .settings-content &gt; h1:not(:first-child)):
* Source/WebInspectorUI/UserInterface/Views/SettingsPopover.js: Copied from Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.css.
(WI.SettingsPopover):
(WI.SettingsPopover.prototype.show):
(WI.SettingsPopover.prototype.set content):
(WI.SettingsPopover.prototype.createContentElement):
(WI.SettingsPopover.prototype._presentOverTargetElement):

`WI.SettingsPopover` handles showing a popover at the correct position on screen,
and wrapping the content provided by subclasses with the `settings-content` CSS
class name which provides basic styling.

Canonical link: <a href="https://commits.webkit.org/260757@main">https://commits.webkit.org/260757@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d8c678b06967417f26ecc95e7a3081bac5076580

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/109274 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/18357 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/42076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/793 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/118497 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/19808 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/85/builds/9649 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/101519 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/115029 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/14848 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/42076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/43027 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/96820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/42076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/84772 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/11131 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/42076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/11851 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/85/builds/9649 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/80/builds/17230 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/42076 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7427 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/13481 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

dcrousso commented on Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
```diff
@@ -167,6 +167,7 @@ localizedStrings["All items in \u0022%s\u0022 must be valid DOM nodes"] = "All i
 /* Label for setting that allows the user to inspect the Web Inspector user interface. */
 localizedStrings["Allow Inspecting Web Inspector @ Experimental Settings"] = "Allow Inspecting Web Inspector";
 localizedStrings["Allow Media Capture on Insecure Sites"] = "Allow Media Capture on Insecure Sites";
+localizedStrings["Allow media capture on insecure sites"] = "Allow media capture on insecure sites";
```
oops?

dcrousso commented on Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
```diff
@@ -526,18 +529,24 @@ localizedStrings["Disable All"] = "Disable All";
 localizedStrings["Disable Event Listener"] = "Disable Event Listener";
 localizedStrings["Disable Event Listeners"] = "Disable Event Listeners";
 localizedStrings["Disable ICE Candidate Restrictions"] = "Disable ICE Candidate Restrictions";
+localizedStrings["Disable ICE candidate restrictions"] = "Disable ICE candidate restrictions";
```
oops?

dcrousso commented on Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
```diff
@@ -1747,6 +1760,7 @@ localizedStrings["Update Local Override"] = "Update Local Override";
 localizedStrings["Usage: %s"] = "Usage: %s";
 localizedStrings["Use Mock Capture Devices"] = "Use Mock Capture Devices";
 localizedStrings["Use default media styles"] = "Use default media styles";
+localizedStrings["Use mock capture devices"] = "Use mock capture devices";
```
oops?

patrickangle replied:
These three strings match the intended capitalization, however it is strange that the previous capitalizations continue to exist.

rcaliman-apple replied:
I had used the script to automatically extract strings. Didn't occur to me it wouldn't also clean-up unused strings. 
Cleaned up manually in the latest patch.

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Main.js:
```diff
@@ -112,6 +112,7 @@ WI.loaded = function()
         WI.browserManager = new WI.BrowserManager,
         WI.targetManager = new WI.TargetManager,
         WI.networkManager = new WI.NetworkManager,
+        WI.deviceSettingsManager = new WI.DeviceSettingsManager,
```
i think our preference has been to add new managers to the end of this list

dcrousso commented on Source/WebInspectorUI/UserInterface/Base/Main.js:
```diff
@@ -2098,393 +2087,18 @@ WI._toggleInspectMode = function(event)
-            if (screenSizeValueInput)
-                return;
+    if (WI._deviceSettingsPopover)
+        return; // Clicking the button while the popover is shown will automatically dismiss the popover. The object is cleared by `WI.didDismissPopover()`.
```
NIT: I kinda feel like we should make it so that clicking the button either dismisses the popover without showing it again (just like what color swatches do) or doesn't dismiss it at all 🤔

patrickangle replied:
Isn't that what this line achieves? Since a click occurs outside the popover, it should clean itself up, and this early return should prevent the popover from presenting itself? Or am I mistaken @rcaliman-apple?

rcaliman-apple replied:
Yes, clicking outside the popover will automatically dismiss the popover [Popover.js#L175-L181](https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Views/Popover.js#L175-L181) but it won't clean itself up.

The popover's internal container (`Popover.element`) gets removed after the fade-out transition occurs [Popover.js#L191](https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Views/Popover.js#L191). Because of this, popovers are ephemeral: once dismissed, they can't be re-shown.

Clicking outside the popover will always dismiss it (unless we explicitly inhibit that, which we don't want to do).

Without some sort of check, clicking on the trigger will dismiss the popover *and* build and present a new one.
We want to keep the behavior of toggling the popover when clicking on the trigger because that's a common behavior in Web Inspector.

Another check we can use is `if (WI._deviceSettingsPopover.element) { return; }` since the `element` will be cleared async after "transitionend", but I think this needlessly exposes implementation details.

The `didDismissPopover` callback on the delegate helps with this. It clears the instance of the popover whose DOM content was just removed to allow for toggling when clicking the trigger.

rcaliman-apple replied:
`InlineSwatch` does the same thing. It's just handled internally: [InlineSwatch.js#L144-L170](https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Views/InlineSwatch.js#L144-L170) and uses `_valueEditor` as the object instance to [check for](https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Views/InlineSwatch.js#L277-L278) and [nullify](https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Views/InlineSwatch.js#L167).

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js:
```diff
@@ -0,0 +1,181 @@
+
+    // Public
+
+    get defaultSettingsChanged() {  return this._overridenDeviceUserAgent || this._overridenDeviceSettings.size > 0; }
```
Style: this is not a simple `get`, so it should be separated onto multiple lines

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js:
```diff
@@ -0,0 +1,181 @@
+
+    get overridenDeviceScreenSize() { return this._overridenDeviceScreenSize; }
+
+    get overridenDeviceSettings() { return this._overridenDeviceSettings; }
```
Style: you can remove the whitespace between simple `get`

Style: please move all simple `get` to be the first things after the `// Public`

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js:
```diff
@@ -0,0 +1,181 @@
+        let target = WI.assumingMainTarget();
+
+        if (!target.hasCommand("Page.overrideSetting"))
+            return;
```
do we need to `callback(...)` here?

ditto with the `return` below

rcaliman-apple replied:
Yes, I added calls to `callback(false)` for early exit and the error state.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js:
```diff
@@ -0,0 +1,181 @@
+            // We've just "disabled" the checkbox, so clear the override instead of applying it.
+            target.PageAgent.overrideSetting(setting, (error) => {
+                if (error) {
+                    console.error(error);
```
`WI.reportInternalError`

ditto below

rcaliman-apple replied:
Good catch!

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js:
```diff
@@ -0,0 +1,181 @@
+        if (!target.hasCommand("Page.overrideSetting"))
+            return;
+
+        if (this._overridenDeviceSettings.has(setting)) {
```
you could avoid repeating a bunch of code using `invoke`
```
let commandArguments = {
    setting,
};

let removed = false;
if (! this._overridenDeviceSettings.has(setting)) {
    removed = true;
    commandArguments.value = value;
}

target.PageAgent.overrideSetting.invoke(commandArguments, (error) => {
    if (error) {
        WI.reportInternalError(error);
        return;
    }

    if (removed)
        this._overriddenDeviceSettings.delete(setting);
    else
        this._overriddenDeviceSettings.set(setting, value);

    callback(removed);
    this.dispatchEventToListeners(WI.DeviceSettingsManager.Event.SettingChanged);
}
```
might even wanna make a utility for `Map.prototype.toggle`

rcaliman-apple replied:
This is a good idea; nice and cleaner. I updated the logic with your example. Instead of `removed`, I opted for `apply` boolean to signal when we want to _apply_ the override and on the reverse we remove it:


```javascript
let apply = false;
if (!this._overridenDeviceSettings.has(setting)) {
    apply = true;
    commandArguments.value = value; // Providing `value` applies the override for `setting`. Not providing it clears the override.
}
let commandArguments = {
    setting,
};

target.PageAgent.overrideSetting.invoke(commandArguments, (error) => {
    if (error) {
        WI.reportInternalError(error);
        callback(false);
        return;
    }

    if (apply)
        this._overridenDeviceSettings.set(setting, value);
    else
        this._overridenDeviceSettings.delete(setting);

    callback(apply);
    this.dispatchEventToListeners(WI.DeviceSettingsManager.Event.SettingChanged);
});
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js:
```diff
@@ -0,0 +1,181 @@
+        if (!target.hasCommand("Page.overrideUserAgent"))
+            return;
+
+        if (!force && (!value || value === "default")) {
```
ditto :78

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js:
```diff
@@ -0,0 +1,181 @@
+
+        let target = WI.assumingMainTarget();
+
+        if (!force && (!value || value === "default")) {
```
ditto :78

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.css:
```diff
@@ -0,0 +1,46 @@
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+.popover .settings-content:is(.device-settings-content) > label {
```
why not just `.settings-content.device-settings-content`?

ditto below

rcaliman-apple replied:
fixed

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js:
```diff
@@ -0,0 +1,324 @@
+
+WI.OverrideDeviceSettingsPopover = class OverrideDeviceSettingsPopover extends WI.SettingsPopover
+{
+    constructor(delegate)
```
you can drop this if nothing else is done besides calling `super` with the same arguments

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js:
```diff
@@ -0,0 +1,324 @@
+
+    _createSettingsSection()
+    {
+        let coreSettingsGroup = {
```
Style: this can be `const` since nothing changes between Web Inspector sessions

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js:
```diff
@@ -0,0 +1,324 @@
+    _createSettingsSection()
+    {
+        let coreSettingsGroup = {
+            name: WI.UIString("Core Features"),
```
IMO this is kinda a weird name.  How about "General" (and `generalSettingsGroup`)?

patrickangle replied:
We resolved in a weekly meeting about these features to use the term "Core Features" for these three fundamental features of the web platform. IMO it helps drive the impact these three feature have, or rather their absence will have. I'm not opposed to "General", but it would be a divergence from our current plan of record.

dcrousso replied:
i dunno IMO the name "Core Features" just feels really wrong

it almost implies that the other things arent "core" which i dont think is all that accurate (e.g. cross origin stuff is pretty core to the web (especially nowadays)).  it also doesnt really match what we've done elsewhere in Web Inspector (e.g. the General pane of the Settings Tab)

rcaliman-apple replied:
I'm not opposed to "General" either, but I'll defer to Patrick who has discussed the details of this redesign with other people.

patrickangle replied:
General settings should typically be things that are relatively safe/non-destructive to toggle. These three settings really aren't that 😅. "Core Features" communicates the impact turning off any of these three settings will have. The web continues to behave as a user would likely expect it to if you disable something like CORS (ya know, minus the security issues). But when you turn one of these off, the web becomes noticeably different, hence the "core"-ness of these three features. However, this isn't a point worth us getting stuck in review for in either direction.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js:
```diff
@@ -0,0 +1,324 @@
+                {name: WI.UIString("Enable Private Click Measurement debug mode"), setting: InspectorBackend.Enum.Page.Setting.PrivateClickMeasurementDebugModeEnabled, value: true},
+
+                // FIXME: Find out whether we need to keep this.
+                // {name: WI.UIString("Ad Click Attribution Debug Mode"), setting: InspectorBackend.Enum.Page.Setting.AdClickAttributionDebugModeEnabled, value: true},
```
why wouldn't we still want this?  even if this is no longer used (and i think it still is?) we'd still wanna show it for when inspecting older platforms

patrickangle replied:
Yeah, we will want to keep this for inspecting older platform. To match the intended capitalization it should be "Ad Click Attribution debug mode"

rcaliman-apple replied:
Got it. I had misunderstood the setting until I re-read the compatibility comment. 
Refactored to add the the settings according to support:

```javascript
// COMPATIBILITY (iOS 14.0): `Page.Setting.AdClickAttributionDebugModeEnabled` was renamed to `Page.Setting.PrivateClickMeasurementDebugModeEnabled`.
if (InspectorBackend.Enum.Page.Setting.PrivateClickMeasurementDebugModeEnabled)
    privacySettingsGroup.settings.push({name: WI.UIString("Enable Private Click Measurement debug mode"), setting: InspectorBackend.Enum.Page.Setting.PrivateClickMeasurementDebugModeEnabled, value: true});

// COMPATIBILITY (iOS 14.0): `Page.Setting.AdClickAttributionDebugModeEnabled` was renamed to `Page.Setting.PrivateClickMeasurementDebugModeEnabled`.
if (InspectorBackend.Enum.Page.Setting.AdClickAttributionDebugModeEnabled)
    privacySettingsGroup.settings.push({name: WI.UIString("Ad Click Attribution debug mode"), setting: InspectorBackend.Enum.Page.Setting.AdClickAttributionDebugModeEnabled, value: true});
```

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js:
```diff
@@ -0,0 +1,324 @@
+                {name: WI.UIString("Use mock capture devices"), setting: InspectorBackend.Enum.Page.Setting.MockCaptureDevicesEnabled, value: true},
+
+                // FIXME: Find out whether we need to keep this.
+                // {name: WI.UIString("Disable Encryption"), setting: InspectorBackend.Enum.Page.Setting.WebRTCEncryptionEnabled, value: false},
```
ditto

patrickangle replied:
This setting is indeed slated to be removed. It's removal had landed in https://commits.webkit.org/258804@main, but it was then reverted in https://commits.webkit.org/258884@main due to a number of new test failures running API tests against the simulator. That mystery has not yet been reinvestigated yet, and may have been complete happenstance.

That said, we should keep the setting here for now until we can tear out the backend hooks for it, otherwise we are much more likely to punt on removing the backend hooks, leaving behind dead code.

rcaliman-apple replied:
Got it. Re-added the "Disable encryption" setting.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js:
```diff
@@ -0,0 +1,324 @@
+            ],
+        };
+
+        const settingsGroups = [coreSettingsGroup];
```
Style: this should be `let` since it's modified below

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js:
```diff
@@ -0,0 +1,324 @@
+
+            userAgentValueInput = userAgentValue.appendChild(document.createElement("input"));
+            userAgentValueInput.spellcheck = false;
+            userAgentValueInput.value = userAgentValueInput.placeholder = WI.deviceSettingsManager.overridenDeviceUserAgent || navigator.userAgent;
```
Aside: hmm i dunno how i never noticed this before, but this wont work for remote inspection.  we might want to do a
```
WI.runtimeManager.evaluateInInspectedWindow(`globalThis.navigator?.userAgent`, {returnByValue: true}, (result, wasThrown) => {
    if (!wasThrown && result)
        this._initialUserAgent = result;
});
```
inside `WI.DeviceSettingsManager.prototype.initializeTarget` 🤔

rcaliman-apple replied:
I understand what the intent is. While it shouldn't technically matter what the free-form user agent field is populated with since the developer is going to replace it with whatever they want anyway, it would indeed be nice to offer the target's actual default user agent as a starting point.

I adapted your example. It needed an `objectGroup` and when using `returnByValue: true` option, the actual result is the third argument in the callback (the first argument is the represented object, which is `null` with this option [RuntimeManager.js#L144-L145](https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Controllers/RuntimeManager.js#L144-L145))

```javascript
const options = {
    objectGroup: WI.DeviceSettingsManager.ObjectGroup,
    returnByValue: true,
}

WI.runtimeManager.evaluateInInspectedWindow("globalThis.navigator?.userAgent", options, (remoteObject, wasThrown, result) => {
    if (!wasThrown && result)
        this._deviceUserAgent = result.value;

    target.RuntimeAgent.releaseObjectGroup(WI.DeviceSettingsManager.ObjectGroup);
});
```

Not sure if `target.RuntimeAgent.releaseObjectGroup()` was necessary, but it seems like good hygiene since this group is used for just this one operation.

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.css:
```diff
@@ -23,29 +23,19 @@
-    display: grid;
-    padding: 5px;
-    gap: 4px 8px;
+.popover .settings-content:is(.user-preferences-content) {
```
why not just `.settings-content.user-preferences-content`?

ditto below

rcaliman-apple replied:
I find that the missing space between classnames is easy to miss when reading stylesheets with our current selector style.

`.user-preferences-content` is essentially a "modifier" for `.settings-content`. In BEM style notation, we'd use a suffix like `--user-preferences` (ex: `.settings-content--user-preferences`) to denote the relationship.

I find that `:is()` helps with this:
- helps communicate the intent of the class (modifier)
- avoids misreading the missing space
- maintains the specificity (not particularly relevant here given our preference for super-precise selectors)

patrickangle replied:
I can appreciate the spoken correctness of saying "settings content that is device settings content", but from a style perspective we have favored simpler selectors, which should theoretically be more performant since they wouldn't need de-suggared. I'd suggest we avoid unnecessary use of `:is` and instead of use it represent situations in which we are deriving a benefit like not having to re-state a selector to change a single class in the middle of the selector.

rcaliman-apple replied:
Done. Removed `:is()`

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/SettingsPopover.css:
```diff
@@ -0,0 +1,41 @@
+.popover .settings-content > h1 {
+    font-size: 11px;
+    margin-block-start: 0;
+    margin-block-end: 4px;
```
```
margin-block: 0 4px;
```

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js:
```diff
@@ -0,0 +1,324 @@
+
+        let userAgentValueSelect = userAgentValue.appendChild(document.createElement("select"));
+
+        const userAgents = [
```
Our patches have crossed in the wind, and this is no longer quite correct. It needs to check the WebKitAdditions global object for user agents, and fall back to these. Double check the current ToT implementation in Main.js

rcaliman-apple replied:
Good catch! I brought over the `WebKitAdditions` check.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.css:
```diff
@@ -23,29 +23,19 @@
+.popover .settings-content:is(.user-preferences-content) > select {
     grid-column: 2;
-}
+}
```
Style: missing newline at end of file

patrickangle commented on Source/WebInspectorUI/UserInterface/Base/Main.js:
```diff
@@ -2098,393 +2087,18 @@ WI._toggleInspectMode = function(event)
 
+WI._handleDeviceSettingChanged = function()
+{
+    WI._deviceSettingsTabBarButton.activated = WI.deviceSettingsManager.defaultSettingsChanged;
```
`defaultSettingsChanged` feels a bit ambiguous to me in this context. My initial reaction was that it was an event name because of how it is phrased. Maybe `hasOverridenDefaultSetting`?

rcaliman-apple replied:
Done. Named `hasOverridenDefaultSettings`

patrickangle commented on Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js:
```diff
@@ -0,0 +1,194 @@
+        if (!this._overridenDeviceSettings.has(setting)) {
+            apply = true;
+            commandArguments.value = value; // Providing `value` applies the override for `setting`. Not providing it clears the override.
+        }
```
Nit:
```
let apply = !this._overridenDeviceSettings.has(setting);
if (apply)
    commandArguments.value = value;
```

Also can we provide a slightly more descriptive name for `apply`? Maybe `shouldApplyValue` or maybe invert the setting and call it `shouldRemoveOverride`? A clearer name would likely eliminate the need for this comment as well. This applies to the methods below as well that have an `apply` used this way.

rcaliman-apple replied:
Sure, we can use a more descriptive name, like `shouldOverride`.

But IMO we should keep the comment. To figure out that _not passing a value_ removes the override isn't immediately obvious. One needs to either read the description of the `overrideSetting` entry in `Page.json` or read through the implementation of `InspectorPageAgent.cpp`.

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js:
```diff
@@ -0,0 +1,321 @@
+
+        // COMPATIBILITY (iOS 14.0): `Page.Setting.AdClickAttributionDebugModeEnabled` was renamed to `Page.Setting.PrivateClickMeasurementDebugModeEnabled`.
+        if (InspectorBackend.Enum.Page.Setting.AdClickAttributionDebugModeEnabled)
+            privacySettingsGroup.settings.push({name: WI.UIString("Ad Click Attribution debug mode"), setting: InspectorBackend.Enum.Page.Setting.AdClickAttributionDebugModeEnabled, value: true});
```
Should the first compatibility comment say that "`Page.Setting.PrivateClickMeasurementDebugModeEnabled` was added"? Or maybe these both share the same COMPATIBILITY comment and the legacy path becomes an `if else (...)`?

dcrousso replied:
yeah our usual approach is to use an `else`

rcaliman-apple replied:
Done. Used `if / else`

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js:
```diff
@@ -0,0 +1,321 @@
+            settingsGroups.push(compatibilitySettingsGroup);
+            settingsGroups.push(securitySettingsGroup);
+            settingsGroups.push(privacySettingsGroup);
+        }
```
I wonder if we should instead create `settingsGroups` at the start and `push` the various groups directly into it conditionally based on `isRemote` to avoid doing the work to create the unused groups for non-remote cases (which is most of the time)? This would also make it clearer at the time the group is declared if it is present for remote inspection or not.

rcaliman-apple replied:
Done

patrickangle commented on Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.js:
```diff
@@ -171,6 +116,47 @@ WI.OverrideUserPreferencesPopover = class OverrideUserPreferencesPopover extends
 
+    // Private
+
+    _createSelectElement({contentElement, id, label, preferenceName, preferenceValues, defaultValue})
```
Nit: `_createSelectElement({contentElement, id, label, preferenceName, preferenceValues, defaultValue} = {})`

dcrousso replied:
IMO that's only necessary if these arguments are all optional.  if they're all always expected then we shouldn't have the `= {}`

tho frankly in that case then these should really be top-level parameters, not properties on a destructured object :/

rcaliman-apple replied:
If there's more than 3-4 arguments, wrapping them in an object makes it easier to compose for the method call regardless of order. I'd keep it as is.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js:
```diff
@@ -0,0 +1,194 @@
+        const options = {
+            objectGroup: WI.DeviceSettingsManager.ObjectGroup,
+            returnByValue: true,
+        }
```
Style: missing `;`

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js:
```diff
@@ -0,0 +1,194 @@
+        }
+
+        const options = {
+            objectGroup: WI.DeviceSettingsManager.ObjectGroup,
```
i'm pretty sure you dont need to provide an `objectGroup` as there's no need to keep the return value alive

also, this is only used in this function, so i'd just inline it and then use it as `options.objectGroup` below

rcaliman-apple replied:
An `objectGroup` is required and asserted for: [RuntimeManager.js#L110](https://github.com/WebKit/WebKit/blob/main/Source/WebInspectorUI/UserInterface/Controllers/RuntimeManager.js#L110).

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js:
```diff
@@ -0,0 +1,194 @@
+        let commandArguments = {};
+
+        let apply = false;
+        if ((value && value !== "default") || (force && value)) {
```
ditto @patrickangle's comment above on :102
```
let shouldOverride = value && (value !== WI.DeviceSettingsManager.DefaultValue || force);
if (shouldOverride)
    commandArguments.value = value;
```

rcaliman-apple replied:
done

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js:
```diff
@@ -0,0 +1,194 @@
+        let commandArguments = {};
+
+        let apply = false;
+        if ((value && value !== "default") || (force && value)) {
```
ditto :133

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js:
```diff
@@ -0,0 +1,194 @@
+                return;
+            }
+
+            if (apply)
```
id use a ternary
```
this._overridenDeviceUserAgent = shouldOverride ? value : null;
```

rcaliman-apple replied:
done

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js:
```diff
@@ -0,0 +1,194 @@
+                return;
+            }
+
+            if (apply)
```
ditto :144

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js:
```diff
@@ -0,0 +1,321 @@
+        let contentElement = document.createElement("div");
+        contentElement.classList.add("device-settings-content");
+
+        // FIXME: webkit.org/b/247809 Enable user agent UI once the UI tracks engine-level changes to the current UA (e.g. via API)
```
IMO if we're gonna have a protocol event like `Page.userAgentChanged` we might just wanna use that instead of trying to fetch it above using `WI.runtimeManager.evaluateInInspectedWindow` (i.e. have it dispatch inside `InspectorPageAgent::enable`)

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js:
```diff
@@ -0,0 +1,321 @@
+                {name: `Firefox ${emDash} Windows`, value: "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:101.0) Gecko/20100101 Firefox/101.0"},
+            ],
+            [
+                {name: WI.UIString("Other"), value: "other"},
```
NIT: we might wanna make this into a constant like `WI.OverrideDeviceSettingsPopover._otherValue`

rcaliman-apple replied:
Done

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js:
```diff
@@ -0,0 +1,321 @@
+                {name: WI.UIString("720p"), value: "1280x720"},
+            ],
+            [
+                {name: WI.UIString("Other"), value: "other"},
```
ditto :175

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js:
```diff
@@ -0,0 +1,321 @@
+        if (selectedOptionElement)
+            userAgentValueSelect.value = selectedOptionElement.value;
+        else if (WI.deviceSettingsManager.overridenDeviceUserAgent) {
+            userAgentValueSelect.value = "other";
```
ditto :175

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js:
```diff
@@ -0,0 +1,321 @@
+
+        userAgentValueSelect.addEventListener("change", () => {
+            let value = userAgentValueSelect.value;
+            if (value === "other") {
```
ditto :175

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js:
```diff
@@ -0,0 +1,321 @@
+        if (selectedScreenSizeOptionElement)
+            screenSizeValueSelect.value = selectedScreenSizeOptionElement.value;
+        else if (WI.deviceSettingsManager.overridenDeviceScreenSize) {
+            screenSizeValueSelect.value = "other";
```
ditto :175

dcrousso commented on Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js:
```diff
@@ -0,0 +1,321 @@
+
+        screenSizeValueSelect.addEventListener("change", () => {
+            let value = screenSizeValueSelect.value;
+            if (value === "other") {
```
ditto :175

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js:
```diff
@@ -0,0 +1,178 @@
+                target.PageAgent.overrideSetting(setting, value);
+        }
+
+        WI.runtimeManager.evaluateInInspectedWindow("globalThis.navigator?.userAgent", {objectGroup: WI.DeviceSettingsManager.ObjectGroup, returnByValue: true}, (remoteObject, wasThrown, result) => {
```
NIT: I'd just make this into a
```
const objectGroup = "user-agent";
```
since it's not really used anywhere else

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js:
```diff
@@ -0,0 +1,178 @@
+    get overridenDeviceScreenSize() { return this._overridenDeviceScreenSize; }
+    get overridenDeviceSettings() { return this._overridenDeviceSettings; }
+
+    get hasOverridenDefaultSettings() {
```
Style: `{` should be on separate line

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js:
```diff
@@ -0,0 +1,178 @@
+        };
+
+        let shouldOverride = !this._overridenDeviceSettings.has(setting)
+            commandArguments.value = value; // Providing `value` applies the override for `setting`. Not providing it clears the override.
```
i think you're missing a `if (shouldOverride)`

NIT: this comment doesn't really add anything as it's a pretty common pattern within Web Inspector where not providing one of the arguments acts as a clear (e.g.

rcaliman-apple replied:
>i think you're missing a if (shouldOverride)

Good catch! 

>NIT: this comment doesn't really add anything

Fine, I'll remove the comment.

dcrousso commented on Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js:
```diff
@@ -0,0 +1,178 @@
+
+        let shouldOverride = value && (value !== WI.DeviceSettingsManager.DefaultValue || force);
+        if (shouldOverride) {
+            let tokens = value.split("x");
```
NIT: you could `let [width, height] = value.split("x");` instead of grabbing them later

rcaliman-apple replied:
Neat!

rcaliman-apple committed:
Web Inspector: Reorganize Web Inspector-scoped device settings
https://bugs.webkit.org/show_bug.cgi?id=252440

Reviewed by Devin Rousso.

Reorganizes the device settings popover to show settings appropriate
to the inspected target. Reuses styles from the popover for user preferences overrides.

Settings are copied verbatim from `Main.js`, organized into groups,
and moved to the new `WI.OverrideDeviceSettingsPopover`. Labels are updated.

A new `WI.DeviceSettingsManager` is introduced to manage the collection of
device settings, user agent, and screen size overrides, previously kept in `Main.js`.
It also holds the responsibility to apply overrides on target initalization.

The methods to set overrides are moved mostly unchanged
from `Main.js` to WI.DeviceSettingsManager`:

```
function applyOverriddenUserAgent() -> DeviceSettingsManager.overrideUserAgent()
function applyOverriddenSetting()   -> DeviceSettingsManager.overrideSetting()
function applyOverriddenScreenSize() -> DeviceSettingsManager.overrideScreenSize()
```

A generalized `WI.SettingsPopover` is extracted from `WI.OverrideUserPreferencesPopover`
to handle showing and positioning the popover. The content is provided by subclasses.

A new `WI.OverrideDeviceSettingsPopover` is introduced to show settings
appropriate to the inspected target. More settings are shown in the popover
for remote targets. For local inspection, most of these settings are available
in the develop menu.

The capability to override screen sizes is available only for non-Cocoa platforms.
The code for setting this override is copied verbatim and adapted slightly to
use the new `WI.DeviceSettingsManager`.

* Source/WebInspectorUI/Localizations/en.lproj/localizedStrings.js:
* Source/WebInspectorUI/UserInterface/Base/Main.js:
(WI.loaded):
(WI.initializeTarget):

Most of the code from `Main.js` has moved to `WI.DeviceSettingsManager`
and `WI.OverrideDeviceSettingsPopover` according to responsibilites (controller or view).

* Source/WebInspectorUI/UserInterface/Controllers/DeviceSettingsManager.js: Added.
(WI.DeviceSettingsManager):
(WI.DeviceSettingsManager.prototype.initializeTarget):
(WI.DeviceSettingsManager.supportsSetScreenSizeOverride):
(WI.DeviceSettingsManager.prototype.get defaultSettingsChanged):
(WI.DeviceSettingsManager.prototype.get overridenDeviceUserAgent):
(WI.DeviceSettingsManager.prototype.get overridenDeviceScreenSize):
(WI.DeviceSettingsManager.prototype.get overridenDeviceSettings):
(WI.DeviceSettingsManager.prototype.overrideDeviceSetting):
(WI.DeviceSettingsManager.prototype.overrideUserAgent):
(WI.DeviceSettingsManager.prototype.overrideScreenSize):

* Source/WebInspectorUI/UserInterface/Main.html:
* Source/WebInspectorUI/UserInterface/Test.html:
* Source/WebInspectorUI/UserInterface/Views/Main.css:
(.device-settings-content): Deleted.
(.device-settings-content > table): Deleted.
(.device-settings-content > table > tr > td:first-child): Deleted.
(.device-settings-content .container): Deleted.
(.device-settings-content .container > * + *): Deleted.
(.device-settings-content .column): Deleted.
(.device-settings-content .user-agent select): Deleted.
(.device-settings-content .user-agent input): Deleted.
(.device-settings-content label + label): Deleted.
(.device-settings-content label > input): Deleted.
(.device-settings-content > .reference-page-link-container): Deleted.

Removed styles for device settings popover from `Main.css`
and replaced with streamlined styles in `SettingsPopover.css`
and `OverrideDeviceSettingsPopover.css`.

* Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.css: Copied from Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.css.
(.popover .settings-content:is(.device-settings-content) > label):
(.popover .settings-content:is(.device-settings-content) > label > input):
(.popover .settings-content:is(.device-settings-content) > .select-with-input-container):
(.popover .settings-content:is(.device-settings-content) > .reference-page-link-container):
* Source/WebInspectorUI/UserInterface/Views/OverrideDeviceSettingsPopover.js: Added.
(WI.OverrideDeviceSettingsPopover):
(WI.OverrideDeviceSettingsPopover.prototype.createContentElement):
(WI.OverrideDeviceSettingsPopover.prototype._createSettingsSection):
(WI.OverrideDeviceSettingsPopover.prototype._createUserAgentSection):
(WI.OverrideDeviceSettingsPopover.prototype._createScreenSizeSection):
* Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.css:
(.popover .settings-content:is(.user-preferences-content)):
(.popover .settings-content:is(.user-preferences-content) > h1):
(.popover .settings-content:is(.user-preferences-content) > label):
(.popover .settings-content:is(.user-preferences-content) > select):
(.popover .user-preferences-content): Deleted.
(.popover .user-preferences-content > h1): Deleted.
(.popover .user-preferences-content > h1:not(:first-child)): Deleted.
(.popover .user-preferences-content > label): Deleted.
(.popover .user-preferences-content > select): Deleted.

Generalized styles for settings popovers in `SettingsPopover.css`
and adapted `OverrideDeviceSettingsPopover.css` to build on them.

* Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.js:
(WI.OverrideUserPreferencesPopover):
(WI.OverrideUserPreferencesPopover.prototype.show):
(WI.OverrideUserPreferencesPopover.prototype._createSelectElement):
(WI.OverrideUserPreferencesPopover.prototype._presentOverTargetElement): Deleted.
(WI.OverrideUserPreferencesPopover.prototype._createContentElement): Deleted.
* Source/WebInspectorUI/UserInterface/Views/SettingsPopover.css: Copied from Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.css.
(.popover .settings-content):
(.popover .settings-content > h1):
(.popover .settings-content > h1:not(:first-child)):
* Source/WebInspectorUI/UserInterface/Views/SettingsPopover.js: Copied from Source/WebInspectorUI/UserInterface/Views/OverrideUserPreferencesPopover.css.
(WI.SettingsPopover):
(WI.SettingsPopover.prototype.show):
(WI.SettingsPopover.prototype.set content):
(WI.SettingsPopover.prototype.createContentElement):
(WI.SettingsPopover.prototype._presentOverTargetElement):

`WI.SettingsPopover` handles showing a popover at the correct position on screen,
and wrapping the content provided by subclasses with the `settings-content` CSS
class name which provides basic styling.

Canonical link: https://commits.webkit.org/260757@main
